################################################################################
SHORT DESCRIPTION: 
################################################################################
XSA-156 (http://xenbits.xen.org/xsa/advisory-156.html)
x86/HVM: always intercept #AC and #DB

################################################################################
LONG DESCRIPTION: 
################################################################################
Source: http://xenbits.xen.org/xsa/advisory-156.html
Patches: xsa156.patch

Both being benign exceptions, and both being possible to get triggered by
exception delivery, this is required to prevent a guest from locking up a CPU
(resulting from no other VM exits occurring once getting into such a loop).

The specific scenarios:
1) #AC may be raised during exception delivery if the handler is set to be a
   ring-3 one by a 32-bit guest, and the stack is misaligned.

2) #DB may be raised during exception delivery when a breakpoint got placed on
a data structure involved in delivering the exception. This can result in
an endless loop when a 64-bit guest uses a non-zero IST for the vector 1 IDT
entry, but even without use of IST the time it takes until a contributory
fault would get raised (results depending on the handler) may be quite long.

################################################################################
CHANGELOG 
################################################################################
Added in OpenXT, Xen 4.3.4 patch-queue: Eric Chanudet <chanudete@ainfosec.com>

################################################################################
PATCHES 
################################################################################
Index: xen-4.3.4/xen/arch/x86/hvm/svm/svm.c
===================================================================
--- xen-4.3.4.orig/xen/arch/x86/hvm/svm/svm.c	2015-11-19 20:01:32.089190471 +0100
+++ xen-4.3.4/xen/arch/x86/hvm/svm/svm.c	2015-11-19 20:01:33.615838720 +0100
@@ -941,10 +941,11 @@
         unlikely(v->arch.hvm_vcpu.debug_state_latch != debug_state) )
     {
         uint32_t intercepts = vmcb_get_exception_intercepts(vmcb);
-        uint32_t mask = (1U << TRAP_debug) | (1U << TRAP_int3);
+
         v->arch.hvm_vcpu.debug_state_latch = debug_state;
         vmcb_set_exception_intercepts(
-            vmcb, debug_state ? (intercepts | mask) : (intercepts & ~mask));
+            vmcb, debug_state ? (intercepts | (1U << TRAP_int3))
+                              : (intercepts & ~(1U << TRAP_int3)));
     }
 
     if ( v->arch.hvm_svm.launch_core != smp_processor_id() )
@@ -2242,8 +2243,9 @@
 
     case VMEXIT_EXCEPTION_DB:
         if ( !v->domain->debugger_attached )
-            goto unexpected_exit_type;
-        domain_pause_for_debugger();
+            hvm_inject_hw_exception(TRAP_debug, HVM_DELIVER_NO_ERROR_CODE);
+        else
+            domain_pause_for_debugger();
         break;
 
     case VMEXIT_EXCEPTION_BP:
@@ -2291,6 +2293,11 @@
         break;
     }
 
+    case VMEXIT_EXCEPTION_AC:
+        HVMTRACE_1D(TRAP, TRAP_alignment_check);
+        hvm_inject_hw_exception(TRAP_alignment_check, vmcb->exitinfo1);
+        break;
+
     case VMEXIT_EXCEPTION_UD:
         svm_vmexit_ud_intercept(regs);
         break;
Index: xen-4.3.4/xen/arch/x86/hvm/vmx/vmx.c
===================================================================
--- xen-4.3.4.orig/xen/arch/x86/hvm/vmx/vmx.c	2015-11-19 20:01:32.102523643 +0100
+++ xen-4.3.4/xen/arch/x86/hvm/vmx/vmx.c	2015-11-19 20:01:33.792503255 +0100
@@ -1122,18 +1122,12 @@
 
 void vmx_update_debug_state(struct vcpu *v)
 {
-    unsigned long mask;
-
     ASSERT(v == current);
 
-    mask = 1u << TRAP_int3;
-    if ( !cpu_has_monitor_trap_flag )
-        mask |= 1u << TRAP_debug;
-
     if ( v->arch.hvm_vcpu.debug_state_latch )
-        v->arch.hvm_vmx.exception_bitmap |= mask;
+        v->arch.hvm_vmx.exception_bitmap |= 1U << TRAP_int3;
     else
-        v->arch.hvm_vmx.exception_bitmap &= ~mask;
+        v->arch.hvm_vmx.exception_bitmap &= ~(1U << TRAP_int3);
     vmx_update_exception_bitmap(v);
 }
 
@@ -2636,9 +2630,10 @@
             exit_qualification = __vmread(EXIT_QUALIFICATION);
             HVMTRACE_1D(TRAP_DEBUG, exit_qualification);
             write_debugreg(6, exit_qualification | 0xffff0ff0);
-            if ( !v->domain->debugger_attached || cpu_has_monitor_trap_flag )
-                goto exit_and_crash;
-            domain_pause_for_debugger();
+            if ( !v->domain->debugger_attached )
+                hvm_inject_hw_exception(vector, HVM_DELIVER_NO_ERROR_CODE);
+            else
+                domain_pause_for_debugger();
             break;
         case TRAP_int3: 
         {
@@ -2699,6 +2694,11 @@
 
             hvm_inject_page_fault(regs->error_code, exit_qualification);
             break;
+        case TRAP_alignment_check:
+            HVMTRACE_1D(TRAP, vector);
+            hvm_inject_hw_exception(vector,
+                                    __vmread(VM_EXIT_INTR_ERROR_CODE));
+            break;
         case TRAP_nmi:
             if ( (intr_info & INTR_INFO_INTR_TYPE_MASK) !=
                  (X86_EVENTTYPE_NMI << 8) )
Index: xen-4.3.4/xen/include/asm-x86/hvm/hvm.h
===================================================================
--- xen-4.3.4.orig/xen/include/asm-x86/hvm/hvm.h	2015-11-19 20:01:32.119190109 +0100
+++ xen-4.3.4/xen/include/asm-x86/hvm/hvm.h	2015-11-19 20:01:33.875835584 +0100
@@ -389,7 +389,10 @@
 })
 
 /* These exceptions must always be intercepted. */
-#define HVM_TRAP_MASK ((1U << TRAP_machine_check) | (1U << TRAP_invalid_op))
+#define HVM_TRAP_MASK ((1U << TRAP_debug)           | \
+                       (1U << TRAP_invalid_op)      | \
+                       (1U << TRAP_alignment_check) | \
+                       (1U << TRAP_machine_check))
 
 /*
  * x86 event types. This enumeration is valid for:
